### ISSUE TITLE
loss and val_loss NAN while implementing Siamese Network using keras model  

### ISSUE THREAD

**Issue by nazish07** [keras-team/keras#12556]
I am trying to implement a siamese network using keras model to develop Sketch Based Image Retrieval using dataset containing images and corresponding sketches.
The features are properly assembled in the form of X (containing batch of images and sketches) and Y (labels) with specifications as:
X.shape - (3300, 2,  1, 100, 100) --[3300 samples, 2-images & sketches, (100,100)-image size]
Y.shape - (3300,1) --[3300 labels (1/0) indicating similar and non-similar samples]

The code is as follows:
import numpy as np
from PIL import Image
import pickle

from sklearn.model_selection import train_test_split
from keras import backend as K
from keras.layers import Activation 
from keras.layers import Input, Lambda, Dense, Dropout, Conv2D, MaxPooling2D, Flatten
from keras.models import Sequential, Model
from keras.optimizers import RMSprop
from keras.layers.normalization import BatchNormalization

import cv2 as cv
import CreateDataset as cdt

X, Y = cdt.get_dataset(50,(100,100))

x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.2)

def build_base_network(input_shape):
    seq = Sequential()
    #nb_filter = [6, 12]
    #kernel_size = 3
    # convolutional layer 1
    seq.add(Conv2D(32,(15,15), strides=1,
                          input_shape = input_shape, padding = 'valid',
                          data_format ='channels_first'))
    seq.add(Activation('relu'))
    seq.add(BatchNormalization())
    seq.add(MaxPooling2D(pool_size = (2,2),strides=2, padding='valid',data_format ='channels_first'))
    seq.add(Dropout(.5))
    
    # convolutional layer 2
    seq.add(Conv2D(64,(8,8), strides=1,
                           padding = 'valid', data_format ='channels_first'))
    seq.add(Activation('relu'))
    seq.add(BatchNormalization())
    seq.add(MaxPooling2D(pool_size=(3,3), strides =3, padding='valid',data_format ='channels_first'))
    seq.add(Dropout(.5))
    
    # convolutional layer 3
    seq.add(Conv2D(256,(5,5), strides=1,
                           padding = 'valid', data_format ='channels_first'))
    seq.add(Activation('relu'))
    seq.add(BatchNormalization())
    seq.add(MaxPooling2D(pool_size=(2,2), strides =2, padding='valid',data_format ='channels_first'))
    seq.add(Dropout(.5))

    #flatten
    seq.add(Flatten()) 
    seq.add(Dense(64, activation = 'relu'))
    seq.add(Dropout(0.5))
    #seq.add(Dense(50, activation = 'relu'))
    
    return seq

input_dim = x_train.shape[2:]
img_a = Input(shape = input_dim)
img_b = Input(shape = input_dim)

base_network = build_base_network(input_dim)
feat_vecs_a = base_network(img_a)
feat_vecs_b = base_network(img_b)

def euclidean_distance(vects):
    x, y =vects
    return K.sqrt(K.sum(K.square(x-y),axis=1, keepdims = True))

def eucl_dist_output_shape(shapes):
    shape1, shape2 = shapes
    return (shape1[0],1)

distance = Lambda(euclidean_distance, 
                  output_shape = eucl_dist_output_shape)([feat_vecs_a,feat_vecs_b])

epochs = 13
rms =RMSprop()
model = Model(input=[img_a, img_b], output = distance)

def contrastive_loss(y_true, y_pred):
    margin = 1
    return K.mean(y_true * K.square(y_pred) + (1 - y_true) *
                  K.square(K.maximum(margin - y_pred, 0)))

model.compile(loss = contrastive_loss, optimizer = rms)

img_1 = x_train[:,0]
img_2 = x_train[:,1]

model.fit([img_1, img_2], y_train, validation_split = .25,batch_size=32,
          verbose = 2, nb_epoch = epochs)

After running this the output is :

Train on 1980 samples, validate on 660 samples
Epoch 1/13
 - 765s - loss: nan - val_loss: nan
Epoch 2/13
 - 781s - loss: nan - val_loss: nan
Epoch 3/13......

I've tried many changes like, changing batch_size, increasing no. of training samples, introducing BatchNormalisation etc. but didn't change output. I've checked the input also and found not a single element to be NAN. Values of all of them lies between 0 and 1. Please guide where the problem could lie.
